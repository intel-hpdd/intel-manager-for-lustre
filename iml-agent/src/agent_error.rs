// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use iml_cmd::CmdError;
use iml_fs::ImlFsError;
use iml_wire_types::PluginName;
use std::{fmt, process::Output};
use thiserror::Error;
use tokio_util::codec::LinesCodecError;

pub type Result<T> = std::result::Result<T, ImlAgentError>;

#[derive(Debug)]
pub struct NoSessionError(pub PluginName);

impl fmt::Display for NoSessionError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "No session found for {:?}", self.0)
    }
}

impl std::error::Error for NoSessionError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct NoPluginError(pub PluginName);

impl fmt::Display for NoPluginError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Plugin in registry not found for {:?}", self.0)
    }
}

impl std::error::Error for NoPluginError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct RequiredError(pub String);

impl fmt::Display for RequiredError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{:?}", self.0)
    }
}

impl std::error::Error for RequiredError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug)]
pub struct CibError(pub String);

impl fmt::Display for CibError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl std::error::Error for CibError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        None
    }
}

#[derive(Debug, Error)]
pub enum ImlAgentError {
    #[error(transparent)]
    AddrParseError(#[from] std::net::AddrParseError),
    #[error(transparent)]
    IpNetworkError(#[from] ipnetwork::IpNetworkError),
    #[error(transparent)]
    CibError(#[from] CibError),
    #[error(transparent)]
    CmdError(#[from] CmdError),
    #[error(transparent)]
    FmtError(#[from] strfmt::FmtError),
    #[error(transparent)]
    FromUtf8Error(#[from] std::string::FromUtf8Error),
    #[error(transparent)]
    ImlFsError(#[from] ImlFsError),
    #[error(transparent)]
    InvalidHeaderValue(#[from] http::header::InvalidHeaderValue),
    #[error(transparent)]
    InvalidUri(#[from] http::uri::InvalidUri),
    #[error(transparent)]
    InvalidUriParts(#[from] http::uri::InvalidUriParts),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    LiblustreError(#[from] liblustreapi::error::LiblustreError),
    #[error(transparent)]
    LinesCodecError(#[from] LinesCodecError),
    #[error(transparent)]
    LustreCollectorError(#[from] lustre_collector::error::LustreCollectorError),
    #[error("Marker Not Found")]
    MarkerNotFound,
    #[error("Argument '{0}' Missing")]
    MissingArgument(String),
    #[error(transparent)]
    NoPluginError(#[from] NoPluginError),
    #[error(transparent)]
    NoSessionError(#[from] NoSessionError),
    #[error(transparent)]
    OneshotCanceled(#[from] futures::channel::oneshot::Canceled),
    #[error(transparent)]
    ParseBoolError(#[from] std::str::ParseBoolError),
    #[error(transparent)]
    ParseIntError(#[from] std::num::ParseIntError),
    #[error(transparent)]
    RequiredError(#[from] RequiredError),
    #[error(transparent)]
    Reqwest(#[from] reqwest::Error),
    #[error("Rx went away")]
    SendError,
    #[error(transparent)]
    Serde(#[from] serde_json::Error),
    #[error(transparent)]
    SystemdError(#[from] iml_systemd::SystemdError),
    #[error(transparent)]
    TokioJoinError(#[from] tokio::task::JoinError),
    #[error(transparent)]
    TokioTimerError(#[from] tokio::time::Error),
    #[error("Unexpected status code")]
    UnexpectedStatusError,
    #[error(transparent)]
    UrlParseError(#[from] url::ParseError),
    #[error(transparent)]
    Utf8Error(#[from] std::str::Utf8Error),
    #[error(transparent)]
    XmlError(#[from] elementtree::Error),
    #[error(transparent)]
    QuickXmlError(#[from] quick_xml::Error),
    #[error("{0}")]
    LdevEntriesError(String),
    #[error(transparent)]
    CombineEasyError(#[from] combine::stream::easy::Errors<char, String, usize>),
}

impl From<Output> for ImlAgentError {
    fn from(output: Output) -> Self {
        ImlAgentError::CmdError(output.into())
    }
}

impl From<futures::channel::mpsc::SendError> for ImlAgentError {
    fn from(_: futures::channel::mpsc::SendError) -> Self {
        ImlAgentError::SendError
    }
}

impl serde::Serialize for ImlAgentError {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        serializer.serialize_str(&format!("{:?}", self))
    }
}
