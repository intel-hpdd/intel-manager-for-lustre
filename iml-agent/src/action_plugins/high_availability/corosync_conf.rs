// Copyright (c) 2020 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

use crate::agent_error::ImlAgentError;
use iml_cmd::{CheckedCommandExt, Command};
use lazy_static::lazy_static;
use regex::Regex;
use std::{collections::HashMap, convert::TryFrom, io};
use tokio::{fs::OpenOptions, io::AsyncWriteExt};

static COROSYNC_CONF: &str = "/etc/corosync/corosync.conf";

/// Reads the corosync configuration.
async fn read_corosync_conf() -> Result<Vec<u8>, io::Error> {
    iml_fs::read_file_to_end(COROSYNC_CONF).await
}

/// Truncates and writes a new configuration.
async fn write_corosync_conf(xs: &[u8]) -> Result<(), ImlAgentError> {
    OpenOptions::new()
        .write(true)
        .truncate(true)
        .open(COROSYNC_CONF)
        .await?
        .write_all(xs)
        .await?;

    Ok(())
}

/// Updates the config_version. This is useful so nodes can tell which conf is newer.
fn update_config_version(
    s: impl Iterator<Item = String>,
    next_version: u64,
) -> impl Iterator<Item = String> {
    s.filter(|x| !x.trim().starts_with("config_version:"))
        .map(move |x| {
            if x.trim().starts_with("version:") {
                format!("{}\n    config_version: {}", x, next_version)
            } else {
                x
            }
        })
}

/// Updates the mcastport for all interfaces in the corosync.conf.
fn update_mcast_port(
    s: impl Iterator<Item = String>,
    new_port: u16,
) -> impl Iterator<Item = String> {
    lazy_static! {
        static ref RE: Regex = Regex::new(r"(\s*mcastport:\s*)\d+").unwrap();
    }

    s.map(move |x| {
        if x.trim().starts_with("mcastport:") {
            RE.replace(&x, |caps: &regex::Captures| {
                format!("{}{}", &caps[1], new_port)
            })
            .to_string()
        } else {
            x
        }
    })
}

/// Given a new mcast port, updates the port on each interface,
/// bumps the config_version and writes out the new corosync.conf.
pub(crate) async fn change_mcast_port(new_port: u16) -> Result<(), ImlAgentError> {
    let conf = read_corosync_conf().await?;
    let conf = std::str::from_utf8(&conf)?
        .trim()
        .lines()
        .map(|x| x.to_string());

    let (totem, _) = corosync_cmapctl().await?;

    let conf = update_config_version(conf, totem.next_version());
    let conf = update_mcast_port(conf, new_port);
    let conf = conf.collect::<Vec<_>>().join("\n");

    write_corosync_conf(conf.as_bytes()).await?;

    Ok(())
}

type StringMap<'a> = HashMap<&'a str, &'a str>;

#[derive(Debug)]
pub(crate) struct Totem {
    config_version: Option<u64>,
    cluster_name: String,
    token: u32,
}

impl Totem {
    fn next_version(&self) -> u64 {
        match self.config_version {
            Some(x) => x + 1,
            None => 1,
        }
    }
}

impl<'a> TryFrom<&'a StringMap<'a>> for Totem {
    type Error = ImlAgentError;

    fn try_from(x: &'a StringMap<'a>) -> Result<Self, Self::Error> {
        Ok(Totem {
            config_version: x
                .get("config_version (u64)")
                .map(|x| x.parse())
                .transpose()?,
            cluster_name: (*try_get(x, "cluster_name (str)")?).to_string(),
            token: try_get(x, "token (u32)")?.parse()?,
        })
    }
}

#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
pub(crate) struct Interface {
    ringnumber: u32,
    bindnetaddr: String,
    mcastaddr: Option<String>,
    broadcast: Option<String>,
    mcastport: u16,
}

impl<'a> TryFrom<(&'a str, StringMap<'a>)> for Interface {
    type Error = ImlAgentError;

    fn try_from((ringnumber, x): (&'a str, StringMap<'a>)) -> Result<Self, Self::Error> {
        Ok(Interface {
            ringnumber: ringnumber.parse()?,
            bindnetaddr: (*try_get(&x, "bindnetaddr (str)")?).to_string(),
            mcastaddr: x.get("mcastaddr (str)").map(|x| (*x).to_string()),
            broadcast: x.get("broadcast (str)").map(|x| (*x).to_string()),
            mcastport: try_get(&x, "mcastport (u16)")?.parse()?,
        })
    }
}

/// View the current totem and interface information from `corosync-cmapctl`.
async fn corosync_cmapctl() -> Result<(Totem, Vec<Interface>), ImlAgentError> {
    let x = Command::new("corosync-cmapctl")
        .arg("totem")
        .kill_on_drop(true)
        .checked_output()
        .await?;

    corosync_cmapctl_parser(&x.stdout)
}

fn corosync_cmapctl_parser(x: &[u8]) -> Result<(Totem, Vec<Interface>), ImlAgentError> {
    let x = std::str::from_utf8(x)?;

    let (totem, interfaces) = x
        .trim()
        .lines()
        .filter_map(|y| {
            let xs: Vec<_> = y.split('=').collect();

            match xs[..] {
                [k, v] => Some((k.trim(), v.trim())),
                _ => None,
            }
        })
        .map(|(k, v)| (k.split('.').collect(), v))
        .fold(
            (HashMap::new(), HashMap::new()),
            |(mut totem, mut interfaces): (StringMap, HashMap<&str, StringMap>),
             (k, v): (Vec<_>, _)| {
                match k[..] {
                    ["totem", y] => {
                        totem.insert(y, v);
                    }
                    ["totem", "interface", x, y] => {
                        interfaces.entry(x).or_default().insert(y, v);
                    }

                    _ => {}
                }

                (totem, interfaces)
            },
        );

    Ok((
        Totem::try_from(&totem)?,
        interfaces
            .into_iter()
            .map(Interface::try_from)
            .collect::<Result<_, _>>()?,
    ))
}

fn try_get<'a>(hm: &'a StringMap<'a>, k: &str) -> Result<&'a &'a str, ImlAgentError> {
    hm.get(k)
        .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, format!("Property {} not found", k)))
        .map_err(|e| e.into())
}

#[cfg(test)]
mod tests {
    use super::{corosync_cmapctl_parser, update_config_version, update_mcast_port};
    use std::iter::FromIterator;

    fn corosync_cmapctl_fixture() -> &'static [u8] {
        include_bytes!("../fixtures/corosync-cmapctl-totem.txt")
    }

    fn corosync_conf_fixture() -> &'static [u8] {
        include_bytes!("../fixtures/corosync.conf")
    }

    #[test]
    fn test_corosync_cmapctl_output() -> Result<(), Box<dyn std::error::Error>> {
        let x = corosync_cmapctl_fixture();

        let mut y = corosync_cmapctl_parser(x)?;

        y.1.sort_unstable();

        insta::assert_debug_snapshot!(y);

        Ok(())
    }

    #[test]
    fn test_update_config_version() -> Result<(), Box<dyn std::error::Error>> {
        let conf = corosync_conf_fixture();

        let conf = std::str::from_utf8(&conf)?
            .trim()
            .lines()
            .map(|x| x.to_string());

        let conf = Vec::from_iter(update_config_version(conf, 10)).join("\n");

        insta::assert_display_snapshot!(conf);

        Ok(())
    }

    #[test]
    fn test_update_mcast_port() -> Result<(), Box<dyn std::error::Error>> {
        let conf = corosync_conf_fixture();

        let conf = std::str::from_utf8(&conf)?
            .trim()
            .lines()
            .map(|x| x.to_string());

        let conf = Vec::from_iter(update_mcast_port(conf, 40001)).join("\n");

        insta::assert_display_snapshot!(conf);

        Ok(())
    }
}
