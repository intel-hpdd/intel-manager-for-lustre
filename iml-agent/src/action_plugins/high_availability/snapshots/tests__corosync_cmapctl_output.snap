---
source: iml-agent/src/action_plugins/high_availability/corosync_conf.rs
expression: y
---
(
    Totem {
        config_version: Some(
            3,
        ),
        cluster_name: "lustre-ha-cluster",
        token: 17000,
    },
    [
        Interface {
            ringnumber: 0,
            bindnetaddr: "10.128.0.22",
            mcastaddr: Some(
                "226.94.0.1",
            ),
            broadcast: None,
            mcastport: 40011,
        },
        Interface {
            ringnumber: 1,
            bindnetaddr: "10.73.10.0",
            mcastaddr: Some(
                "226.94.1.1",
            ),
            broadcast: None,
            mcastport: 40011,
        },
    ],
)
